server/monitor-types/tls.js

```diff
@@ -0,0 +1,266 @@
+                message = "Connection successful";
+            }
+
+            if (success) {
+                this.log("check successful: " + message);
+                heartbeat.msg = message;
+                heartbeat.status = UP;
+            } else {
+                this.log("check failed: " + message);
+                throw new Error(message);
+            }
+        }
+    }
+
+    /**
+     * Sends the request over the given TLS socket and returns the response.
+     * @param {AbortController} aborter   Abort controller used to abort the request
+     * @param {TLSSocket}       tlsSocket TLS socket instance
+     * @param {*}               options   Monitor options
+     * @returns {Promise<string>} Server response on success or rejected promise on error
+     */
+    async getResponseFromTlsPort(aborter, tlsSocket, options) {
+        if (options.request) {
+            this.debug_log(`sending request: '${options.request}'`);
+            tlsSocket.write(options.request);
+        }
+
+        return await this.readData(aborter, tlsSocket);
+    }
+
+    /**
+     * Connects to a given host and port using native TLS or STARTTLS.
+     * @param {AbortController} aborter     Abort controller used to abort the connection
+     * @param {string}          hostname    Host to connect to
+     * @param {int}             port        TCP port to connect to
+     * @param {boolean}         useStartTls True if STARTTLS should be used, false for native TLS
+     * @returns {Promise<TLSSocket>} TLS socket instance if successful or rejected promise on error
+     */
+    async connect(aborter, hostname, port, useStartTls) {
+        if (useStartTls) {
+            const socket = new net.Socket({
+                signal: aborter
+            });
+            socket.connect(port, hostname);
+            this.debug_log("TCP connected");
+
+            await this.startTls(aborter, socket);
+            this.debug_log("STARTTLS prelude done");
+
+            const tlsSocket = await this.upgradeConnection(socket);
+            this.debug_log("TLS upgrade done");
+            tlsSocket.on("close", (hadError) => {
+                socket.end();
+            });
+            return tlsSocket;
+        } else {
+            const tlsSocket = tls.connect(port, hostname, {
+                signal: aborter,
+                servername: hostname
+            });
+            this.debug_log("TLS connected");
+            return tlsSocket;
+        }
+    }
+
+    /**
+     * Reads available data from the given socket.
+     * @param {AbortController} aborter Abort controller used to abort the read
+     * @param {*}               socket  net.Socket or tls.TLSSocket instance to use
+     * @returns {Promise<string>} Data read from the socket or rejected promise on error
+     */
+    readData(aborter, socket) {
+        return new Promise((resolve, reject) => {
+            const cleanup = function () {
+                socket.removeListener("error", onError);
```

Comment:
Could you explain this part (not a js expert here):

Is the event listener being added globally, or locally?
If it is global, that could, right?

Reply:
The event handlers on `aborter` and `socket` are "global" but they're only installed temporarily. They are removed again before the promise settles. So the flow of this function is basically:

1. Install event handlers
2. Resume the socket since we're now ready to receive data
3. Wait for data or error or abort
4. Pause the socket to prevent more data from arriving
5. Uninstall event handlers

I've added some comments now and have also switched the order of pause/uninstall, so the cleanup part reads:
```js
            const cleanup = function () {
                // Pause reading of data (i.e. emission of 'data' events), so that we don't lose
                // any data between now and the next call to readData() while there are no event
                // listeners installed.
                socket.pause();
                socket.removeListener("error", onError);
                socket.removeListener("data", onData);
                aborter.removeEventListener("abort", onAbort);
            };
```